pkg/reconciler/broker/controller.go

```diff
@@ -66,7 +67,11 @@ func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl
 	// Attempt to create a pubsub client for all worker threads to use. If this
 	// fails, pass a nil value to the Reconciler. They will attempt to
 	// create a client on reconcile.
-	client, err := newPubsubClient(ctx, "")
+	projectID, err := utils.ProjectID(os.Getenv(utils.ProjectIDEnvKey), metadataClient.NewDefaultMetadataClient())
+	if err != nil {
+		logging.FromContext(ctx).Fatal("Failed to get project ID", zap.Error(err))
```

Comment:
This will crash the controller container, stopping any controllers from running unless the project id is present. No objects will be reconciled or have statuses updated until the project id is correctly provided, even if those objects don't depend on the configured project id.

Why is that better than the current strategy?

(which is)
- Attempt to get the project id, and if successful create a single client for all worker threads to use
- If client creation fails, defer and attempt to create a client on each reconcile
- For each reconcile that fails to create a client, set an error status on the object being reconciled

Reply:
I reverted the change with a small improvement: don't read projectID again if it's not empty in the Reconciler.

Reply:
I agree we could stand to do more investigation about what this failure really means, and if it's transient or permanent.

We want to eliminate the debug path for the user persona of "resource status isn't saying anything useful, find the controller pod and look at its status/logs". The user persona shouldn't need to be aware of where or how the controller is running. They only care about the resource.

For a concrete example, when was the last time you had to look at the k8s apiserver logs? Might be worth looking into how init failures are handled in storage plugins.